src/server/game/AI/PlayerAI/PlayerAI.cpp

```diff
@@ -94,12 +562,97 @@ bool PlayerAI::IsPlayerRangedAttacker(Player const* who)
             return false;
         }
         case CLASS_PRIEST:
-            return who->HasSpell(SPELL_SHADOWFORM);
+            return (PlayerAI::GetPlayerSpec(who) == SPEC_PRIEST_SHADOW);
         case CLASS_SHAMAN:
-            return !who->HasSpell(SPELL_STORMSTRIKE);
+            return (PlayerAI::GetPlayerSpec(who) == SPEC_SHAMAN_ELEMENTAL);
         case CLASS_DRUID:
-            return who->HasSpell(SPELL_MOONKIN_FORM);
+            return (PlayerAI::GetPlayerSpec(who) == SPEC_DRUID_BALANCE);
+    }
+}
+
+PlayerAI::TargetedSpell PlayerAI::VerifySpellCast(uint32 spellId, Unit* target)
+{
+    // Find highest spell rank that we know
+    uint32 knownRank, nextRank;
+    if (me->HasSpell(spellId))
+        // this will save us some lookups if the player has the highest rank (expected case)
+        nextRank = sSpellMgr->GetNextSpellInChain(knownRank = spellId);
+    else
+        knownRank = 0, nextRank = sSpellMgr->GetFirstSpellInChain(spellId);
+
+    while (nextRank && me->HasSpell(nextRank))
+        nextRank = sSpellMgr->GetNextSpellInChain(knownRank = nextRank);
+
+    if (!knownRank)
+        return {};
+
+    SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(knownRank);
+    if (!spellInfo)
+        return {};
+
+    if (me->GetSpellHistory()->HasGlobalCooldown(spellInfo))
+        return {};
+
+    Spell* spell = new Spell(me, spellInfo, TRIGGERED_NONE);
+    if (spell->CanAutoCast(target))
+        return{ spell, target };
+
+    delete spell;
+    return {};
+}
+
+PlayerAI::TargetedSpell PlayerAI::VerifySpellCast(uint32 spellId, SpellTarget target)
+{
+    Unit* pTarget = nullptr;
+    switch (target)
+    {
+        case TARGET_NONE:
+            break;
+        case TARGET_VICTIM:
+            pTarget = me->GetVictim();
+            if (!pTarget)
+                return {};
+            break;
+        case TARGET_CHARMER:
+            pTarget = me->GetCharmer();
+            if (!pTarget)
+                return {};
+            break;
+        case TARGET_SELF:
+            pTarget = me;
+            break;
+    }
+
+    return VerifySpellCast(spellId, pTarget);
+}
+
+PlayerAI::TargetedSpell PlayerAI::SelectSpellCast(PossibleSpellVector& spells)
```

Comment:
Without the horrible code creating new Spell (and deleting it here), this entire function can be replaced by Trinity::Containers::SelectRandomWeightedContainerElement https://github.com/TrinityCore/TrinityCore/blob/3.3.5/src/common/Utilities/Containers.h#L103